### PR TITLE
DEV-009-Product-Selecting

### PR DESCRIPTION
**Pull request by lenshuygh**
add ProductList caching and Add products to cart

### CODE REVIEW

File: src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java [hunk 1]
```diff
@@ -34,5 +40,37 @@ RestTemplate restTemplate(RestTemplateBuilder builder) {
...
+
+    @Bean
+    public CartDto getCurrentCart() {
+        List<Product> productList = new ArrayList<>();
+        return CartDto.builder()
+                .cartCreationDateTime(LocalDateTime.now())
+                .finalizedCart(false)
+                .activeCart(true)
+                .paidCart(false)
+                .products(productList)
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java:52]: extractie is ok voor leesbaarheid, maar hier mag de productlist gewoon inline -> .products(new ArrayList<>())

File: src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java [hunk 2]
```diff
@@ -34,5 +40,37 @@ RestTemplate restTemplate(RestTemplateBuilder builder) {
...
+                .build();
+    }
+
+    // todo: DEV-015 might not create the SearchModelDto here
+
+    @Bean
+    public SearchDto getSearchModelDto() {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java:59]: rename method to match the returntype -> getSearchDto

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+public class CartController {
+    private DataStore dataStore;
+
+    @Autowired
+    public CartController(DataStore dataStore) {
+        this.dataStore = dataStore;
+    }
+
+    @GetMapping("/details")
+    public String setAfterLoginForUserCart() {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java:22]: rename method to updateStore

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        log.info("setAfterLogintoUserCart");
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java:23]: change info to -> redirect user to secure cartpage (for example)

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/ProductController.java [hunk 5]
```diff
@@ -45,4 +52,17 @@ public String archiveProduct(@PathVariable Long id, Model model) {
         model.addAttribute("status", statusNotification);
         return "product/productDetails";
     }
+
+    @PostMapping
+    public String addSelectedProductToCart(@ModelAttribute("id") Long id) {
+        log.info("chosen product id-> " + id);
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/ProductController.java:58]: conform logger

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java [hunk 6]
```diff
@@ -1,31 +1,58 @@
...
+        //List<Product> productList = productService.findBySearchRequest(searchService.getSearchDto().getBasicSearchTerm()).getProducts();
+        List<Product> productList = productListCacheService.findBySearchRequest(searchService.getSearchDto().getBasicSearchTerm()).getProducts();
         model.addAttribute("products", productList);
+        model.addAttribute("cart", cartService.getCart());
+        return "product/productOverview";
+    }
+
+    @PostMapping
+    public String addSelectedProductToCart(@ModelAttribute("product") Product product, Model model) {
+        log.info("(searchPage) chosen product -> " + product);
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java:48]: conform logger

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java [hunk 7]
```diff
@@ -1,31 +1,58 @@
...
+        //productService.addToCart(product);
+        productService.addToCart(productListCacheService.findById(product.getId()));
+        // todo: (DEV-015) the searchTerm is retrieved from the from the SearchModelDto here, if 'showSearchBarResult' is changed to not use @RequestParam's the code below can be replaced wiht a redirect to '/search' to use the showSearchBarResult method
+        //List<Product> productList = productService.findBySearchRequest(searchService.getSearchDto().getBasicSearchTerm()).getProducts();
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java:52]: remove unnecessary comment

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/UserController.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+@Slf4j
+@Controller
+@RequestMapping("/users")
+public class UserController {
+    private final String CART_DETAIL_TEMPLATE = "redirect:/users/cart/detail";
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/UserController.java:12]: remove final String and inline. Unnecessary final below 3 references

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/UserController.java [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+    @GetMapping("/cart")
+    public String displayCartWhenLoggedIn() {
+        log.info("go to detail");
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/UserController.java:16]: conform logger

File: src/main/java/be/syntra/devshop/DevshopFront/models/ProductListCache.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Setter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@NoArgsConstructor
+@Getter
+@Setter
+public class ProductListCache {
+    List<Product> cachedProductList = new ArrayList<>();
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/models/ProductListCache.java:14]: change name to products

File: src/main/java/be/syntra/devshop/DevshopFront/models/dto/ProductDto.java [hunk 11]
```diff
@@ -30,4 +29,6 @@
     @NotBlank(message = "This is a required field")
     private String description;
     private boolean archived;
+    @Builder.Default
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/models/dto/ProductDto.java:32]: space between fields for readability

File: src/main/java/be/syntra/devshop/DevshopFront/models/dto/SearchDto.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import lombok.*;
+
+@Getter
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class SearchDto {
+    // todo: expand model with DEV-015
+    private String basicSearchTerm;
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/models/dto/SearchDto.java:12]: change variable to searchRequest

File: src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java [hunk 13]
```diff
@@ -38,6 +45,11 @@ protected String determineTargetUrl(Authentication authentication) {
             if (grantedAuthority.getAuthority().equals("ROLE_ADMIN")) {
                 return "/admin/overview";
             } else if (grantedAuthority.getAuthority().equals("ROLE_USER")) {
+                if (dataStore.getMap().get("redirectToCartAfterUserSuccessfulLogin")) {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java:48]: extract to private method -> for example: checkDataStoreForRedirectStrategy

File: src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java [hunk 14]
```diff
@@ -38,6 +45,11 @@ protected String determineTargetUrl(Authentication authentication) {
...
+                    log.info("going after login to cart page");
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java:49]: conform logger

File: src/main/java/be/syntra/devshop/DevshopFront/services/CartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Autowired
+    private CartServiceImpl(CartDto currentCart) {
+        this.currentCart = currentCart;
+    }
+
+    @Autowired
+    private RestTemplate restTemplate;
+
+    //@Cacheable(value = "currentCart")
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/CartServiceImpl.java:29]: remove comment

File: src/main/java/be/syntra/devshop/DevshopFront/services/CartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+        return currentCart;
+    }
+
+    @Override
+    public Product addToCart(Product product) {
+        product.setNumberInCart(product.getNumberInCart() + 1);
+        List<Product> productList = getCart().getProducts();
+        productList.add(product);
+        getCart().setProducts(productList);
+        log.info("product added , new cart size -> " + currentCart.getProducts().size());
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/CartServiceImpl.java:41]: conform logger

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    @Override
+    public void updateProductListCache() {
+        productListCache.setCachedProductList(productService.findAllNonArchived().getProducts());
+    }
+
+    @Override
+    public ProductListCache getProductListCache() {
+        if (dataStore.getMap().get("cacheNeedsUpdate")) {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:34]: extract to private method for readability

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public ProductList findBySearchRequest(String searchRequest) {
+        var result = getProductListCache()
+                .getCachedProductList().stream()
+                .filter(product ->
+                        product.getName()
+                                .toLowerCase()
+                                .contains(searchRequest.toLowerCase()))
+                .collect(Collectors.toList());
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:49]: collect to UnmodifiableList

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        return result.isEmpty() ? new ProductList(getProductListCache().getCachedProductList()) : new ProductList(result);
+    }
+
+    @Override
+    public Product findById(Long id) {
+        Optional<Product> productOptional = getProductListCache().getCachedProductList().stream().filter(product -> product.getId() == id).findFirst();
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:55]: change full methodbody to: return getProductListCache().getCachedProductList().stream.filter(...).ifPresentOrElseThrow(ex.)

File: src/main/java/be/syntra/devshop/DevshopFront/services/SearchService.java [hunk 20]
```diff
@@ -0,0 +1,11 @@
+package be.syntra.devshop.DevshopFront.services;
+
+import be.syntra.devshop.DevshopFront.models.dto.SearchDto;
+
+// todo: replace or alter with DEV-015
+
+public interface SearchService {
+    void captureSearchedTerm(String searchTerm);
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/SearchService.java:8]: searchTerm -> searchRequest

File: src/main/java/be/syntra/devshop/DevshopFront/models/dto/SearchDto.java [hunk 21]
```diff
@@ -0,0 +1,13 @@
+package be.syntra.devshop.DevshopFront.models.dto;
+
+import lombok.*;
+
+@Getter
+@Setter
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class SearchDto {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/models/dto/SearchDto.java:10]: rename to searchModel

File: src/main/java/be/syntra/devshop/DevshopFront/services/SearchServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Autowired
+    public SearchServiceImpl(SearchDto searchDto) {
+        this.searchDto = searchDto;
+    }
+
+    // todo: DEV-015 impl's from interface might replace captureSearchedTerm & getSearchModelDto
+
+    @Override
+    public void captureSearchedTerm(String searchTerm) {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/SearchServiceImpl.java:22]: rename parameter to searchRequest

File: src/main/java/be/syntra/devshop/DevshopFront/services/utils/ProductMapperUtil.java [hunk 23]
```diff
@@ -12,6 +12,7 @@ public static ProductDto convertToProductDto(Product product) {
                 .price(product.getPrice())
                 .description(product.getDescription())
                 .archived(product.isArchived())
+                .numberInCart(product.getNumberInCart())
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/utils/ProductMapperUtil.java:15]: amountInCart?

File: src/main/resources/templates/product/productDetails.html [hunk 24]
```diff
@@ -29,7 +29,15 @@
                         <span class="mdl-list__item-text-body"
                               th:text="'€ ' + ${product.price}">Product Price</span>
                             <a th:if="${product.archived == false}" class="mdl-list__item-secondary-action" href="#">
-                                <em class="material-icons">add_shopping_cart</em></a>
+                                <form method="post" th:object="${product}"
+                                      th:action="@{/products/details/addtocart/{id}(id=${product.id})}">
+                            <input type="hidden" name="id" th:value="${product.id}"/>
+                            <span class="mdl-list__item-secondary-action">
+                                    <button th:disabled="${product.numberInCart > 0}">
+                                        <i class="material-icons">add_shopping_cart</i>
```
Comment (Paul-Gerarts) [inline src/main/resources/templates/product/productDetails.html:37]: change i to em

File: src/test/java/be/syntra/devshop/DevshopFront/controllers/ProductControllerTest.java [hunk 25]
```diff
@@ -103,4 +119,25 @@ void canArchiveProductTest() throws Exception {
...
+        // when
+        final ResultActions getResult = mockMvc.perform(post("/products/")
+                .contentType(MediaType.MULTIPART_FORM_DATA)
+                .param("id", dummyProduct.getId().toString())
+                .param("name", dummyProduct.getName())
+                .param("price", dummyProduct.getPrice().toString()));
+
+        // then
+        getResult
+                .andExpect(status().is3xxRedirection())
```
Comment (Paul-Gerarts) [inline src/test/java/be/syntra/devshop/DevshopFront/controllers/ProductControllerTest.java:137]: change to more specific status

File: src/test/java/be/syntra/devshop/DevshopFront/controllers/SearchControllerTest.java [hunk 26]
```diff
@@ -36,14 +42,31 @@
...
         // given
         final String searchRequest = "product";
         final List<Product> dummyProducts = getDummyNonArchivedProductList();
         final ProductList dummyProductList = new ProductList(dummyProducts);
-        when(productService.findBySearchRequest(searchRequest)).thenReturn(dummyProductList);
+        final CartDto dummyCart = CartUtils.getCartWithOneDummyProduct();
+        final SearchDto dummySearchDto = new SearchDto();
+        dummySearchDto.setBasicSearchTerm(searchRequest);
 
+        //when(productService.findBySearchRequest(any())).thenReturn(dummyProductList);
```
Comment (Paul-Gerarts) [inline src/test/java/be/syntra/devshop/DevshopFront/controllers/SearchControllerTest.java:65]: remove comments behind the when

File: src/test/java/be/syntra/devshop/DevshopFront/controllers/UserControllerTest.java [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+    @Autowired
+    MockMvc mockMvc;
+
+    @Test
+    void displayCartWhenLoggedIn() throws Exception {
+        // when
+        final ResultActions getResult = mockMvc.perform(get("/users/cart"));
+
+        // then
+        getResult.andExpect(status().is3xxRedirection());
```
Comment (Paul-Gerarts) [inline src/test/java/be/syntra/devshop/DevshopFront/controllers/UserControllerTest.java:35]: assert on more specific status

File: src/test/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImplTest.java [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+    @Test
+    void findBySearchRequest() {
+        // given
+        Map<String, Boolean> dummyMap = new HashMap<>();
+        dummyMap.put("cacheNeedsUpdate", true);
+        when(productService.findAllNonArchived()).thenReturn(new ProductList(ProductUtils.getDummyNonArchivedProductList()));
+        when(dataStore.getMap()).thenReturn(dummyMap);
+
+        // when
+        var result = productListCacheService.findBySearchRequest("test");
```
Comment (Paul-Gerarts) [inline src/test/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImplTest.java:80]: make typesafe

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java (hunk 3 above)
Comment (lenshuygh) [reply src/main/java/be/syntra/devshop/DevshopFront/controllers/CartController.java:22]: renamed to 'gotoCartAfterLogin'

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java (hunk 19 above)
Comment (lenshuygh) [reply src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:55]: changed to 
```
...findFirst()
                .orElseThrow(() -> new ProductNotFoundException("...
```

File: src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java (hunk 2 above)
Comment (Paul-Gerarts) [reply src/main/java/be/syntra/devshop/DevshopFront/configuration/WebConfig.java]: Nu de SearchDto een SearchModel is geworden, misschien best de methode wel aanpassen naar getSearchModel

File: src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java [hunk 29]
```diff
@@ -1,31 +1,56 @@
...
+        this.productListCacheService = productListCacheService;
     }
 
     @GetMapping("/")
     public String showSearchBarResult(@RequestParam String searchRequest, Model model) {
-        List<Product> productList = productService.findBySearchRequest(searchRequest).getProducts();
+        // todo: (DEV-015) the search string is saved into the SearchModelDto starting here
+        searchService.captureSearchRequest(searchRequest);
+        // todo: (DEV-015) the searchTerm is retrieved from the from the SearchModelDto here
+        //List<Product> productList = productService.findBySearchRequest(searchService.getSearchDto().getBasicSearchTerm()).getProducts();
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/controllers/SearchController.java:39]: remove unnecessary comment

File: src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java [hunk 30]
```diff
@@ -46,6 +58,14 @@ protected String determineTargetUrl(Authentication authentication) {
         return "auth/login";
     }
 
+    private Boolean disableDataStoreRedirectStrategyAfterLogin() {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/security/AuthSuccessHandler.java:61]: switch both private methods to primitive boolean

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java [hunk 31]
```diff
@@ -0,0 +1,68 @@
...
+    @Override
+    public ProductListCache getProductListCache() {
+        if (checkIfProductsCacheNeedsUpdate()) {
+            disableProductsCacheNeedsUpdate();
+            updateProductListCache();
+        }
+        return productListCache;
+    }
+
+    private Boolean disableProductsCacheNeedsUpdate() {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:40]: more desciptive name would be: updateCacheToFalse

File: src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java [hunk 32]
```diff
@@ -0,0 +1,68 @@
...
+        return dataStore.getMap().put("cacheNeedsUpdate", false);
+    }
+
+    private Boolean checkIfProductsCacheNeedsUpdate() {
```
Comment (Paul-Gerarts) [inline src/main/java/be/syntra/devshop/DevshopFront/services/ProductListCacheServiceImpl.java:44]: both private methods should change returntype to primitive boolean

**Pull request merged**